test/expect.js

```diff
@@ -602,6 +602,16 @@ describe('expect', function () {
     expect({ foo: 1, bar: 2 }).to.not.have.all.keys(['baz', 'foo']);
     expect({ foo: 1, bar: 2 }).to.not.contain.all.keys(['baz', 'foo']);
 
+    // issues/359
+    (function fix359() {
+      var expected = [ 'b', 'a' ];
+      var backup = expected.slice(0);
+      var obj = { "b": 1, "a" :1 };
+      expect(expected).deep.equal(backup);
+      expect(obj).keys(expected);
+      expect(expected).deep.equal(backup);
+    })()
+
```

Comment:
0.  Thanks for the compliments!  `Chai` is super useful on my most recent project at Mozilla.

1.  Closure it because 
    - it seemed a little gross no matter where I put it.  
    - I don't want to leak any of the vars out, and want to 
    - make it clear it's part of  `keys(array)` testing.

2.  Nested 'it' isn't called, afaict (so sad!), so tried this:

```javascript
  it.only("keys(array), array should keep original order (#359)", function () {
    var expected = [ 'b', "a" ];
    var backup = expected.slice(0);
    var obj = { "b": 1, "a" :1 };
    expect(expected).deep.equal(backup);
    expect(obj).keys(expected);
    expect(expected).deep.equal(backup);
  });
```

It's clearer, but less elegant.  

As an inline alternative.  

```javascript
    var issue359 = {
      original:  ['b', 'a'],
      expected: [ 'b', "a" ],
      obj: { "b": 1, "a" : 1 };
    }

    expect(issue359.expected).deep.equal(issue359.original);
    expect(issue359.obj).keys(issue359.expected);
    expect(issue359.expected, "keys should maintain original order").deep.equal(issue359.backup);

```
Or 

```javascript
    var keys_maintain_original_order_issue359 = function () {
      var expected = [ 'b', 'a' ];
      var original_order = expected.slice(0);
      var obj = { "b": 1, "a" :1 };
      expect(expected).deep.equal(original_order);
      expect(obj).keys(original_order);
      expect(expected, "sorting happened unexpectedly in keys()")
        .deep.equal(original_order);
    }
    keys_maintain_original_order_issue359();
```

Anyway, this is a weird case.  I think I lean toward the last style.  And this is probably more time on this than is really needed :)  (But fun to consider alternatives!)

(pull updated to reflect last)